http-server/src/main/java/com/facebook/airlift/http/server/HttpServerChannelListener.java

```diff
@@ -27,78 +28,85 @@
+        }
     }
 
-    @SuppressWarnings("unchecked")
     @Override
-    public void onRequestEnd(Request request)
+    protected void onResponseWrite(Request request, boolean last, ByteBuffer content)
     {
-        long requestBeginTime = (Long) request.getAttribute(REQUEST_BEGIN_ATTRIBUTE);
-        request.setAttribute(REQUEST_BEGIN_TO_END_ATTRIBUTE, System.nanoTime() - requestBeginTime);
+        if (request.getAttribute(RESPONSE_CONTENT_TIMESTAMPS_ATTRIBUTE) == null) {
+            request.setAttribute(RESPONSE_CONTENT_TIMESTAMPS_ATTRIBUTE, new ArrayList<>());
+        }
+        List<Long> contentTimestamps = (List<Long>) request.getAttribute(RESPONSE_CONTENT_TIMESTAMPS_ATTRIBUTE);
+        contentTimestamps.add(System.nanoTime());
     }
 
     @Override
-    public void onResponseBegin(Request request)
+    protected void onComplete(Request request, @Nullable Throwable failure)
     {
-        if (request.getAttribute(REQUEST_BEGIN_TO_END_ATTRIBUTE) == null) {
-            onRequestEnd(request);
-        }
-        request.setAttribute(RESPONSE_CONTENT_TIMESTAMPS_ATTRIBUTE, new ArrayList<Long>());
+        onCompleteHelper(request, -1, HttpFields.EMPTY, failure);
     }
 
-    @SuppressWarnings("unchecked")
     @Override
-    public void onResponseContent(Request request, ByteBuffer content)
+    protected void onComplete(Request request, int status, HttpFields headers, @Nullable Throwable failure)
     {
-        List<Long> contentTimestamps = (List<Long>) request.getAttribute(RESPONSE_CONTENT_TIMESTAMPS_ATTRIBUTE);
-        contentTimestamps.add(System.nanoTime());
+        onCompleteHelper(request, status, headers, failure);
     }
 
-    @SuppressWarnings("unchecked")
-    @Override
-    public void onComplete(Request request)
+    private void onCompleteHelper(Request request, int status, HttpFields headers, @Nullable Throwable failure)
     {
+        long requestBeginTime = (Long) request.getAttribute(REQUEST_BEFORE_HANDLE);
         List<Long> contentTimestamps = (List<Long>) request.getAttribute(RESPONSE_CONTENT_TIMESTAMPS_ATTRIBUTE);
         long firstToLastContentTimeInMillis = -1;
-        if (contentTimestamps.size() > 0) {
+        if (!contentTimestamps.isEmpty()) {
             firstToLastContentTimeInMillis = NANOSECONDS.toMillis(contentTimestamps.get(contentTimestamps.size() - 1) - contentTimestamps.get(0));
         }
-        long beginToDispatchMillis = NANOSECONDS.toMillis((Long) request.getAttribute(REQUEST_BEGIN_TO_DISPATCH_ATTRIBUTE));
-        long beginToEndMillis = NANOSECONDS.toMillis((Long) request.getAttribute(REQUEST_BEGIN_TO_END_ATTRIBUTE));
+        Long afterHandleMillis = (Long) request.getAttribute(REQUEST_AFTER_HANDLE);
+        if (afterHandleMillis == null) {
+            afterHandleMillis = NANOSECONDS.toMillis(System.nanoTime());
+        }
         logger.log(request,
-                request.getResponse(),
```

Comment:
We used to the plumbed `Response` to track response size. Now that we don't have it anymore, should we track this metric separately, maybe by hooking up with `onResponseWrite` ?

Reply:
This is a good idea. I have implemented a change which tracks the buffer sizes and adds them up before logging the request. I changed the `Response response` parameter to be `long responseSize` instead